pkg/cmd/cli/config/smart_merge.go

```diff
@@ -13,13 +14,75 @@ import (
 	clientcmdapi "github.com/GoogleCloudPlatform/kubernetes/pkg/client/clientcmd/api"
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
 
+	osclient "github.com/openshift/origin/pkg/client"
 	"github.com/openshift/origin/pkg/cmd/flagtypes"
 )
 
 var invalidSafeStringSep = regexp.MustCompile(`[.:/]`)
 
+func getClusterNickname(clientCfg *client.Config) (string, error) {
+	serverAddr := flagtypes.Addr{}
+	if err := serverAddr.Set(clientCfg.Host); err != nil {
+		return "", err
+	}
+
+	// we need a character other than "." to avoid conflicts with
+	return fmt.Sprintf("%v:%v", strings.Replace(serverAddr.Host, ".", "-", -1), serverAddr.Port), nil
+}
+
+func getUserNickname(clientCfg *client.Config) (string, error) {
+	client, err := osclient.New(clientCfg)
+	if err != nil {
+		return "", err
+	}
+	userInfo, err := client.Users().Get("~")
+	if err != nil {
+		return "", err
+	}
+
+	clusterNick, err := getClusterNickname(clientCfg)
+	if err != nil {
+		return "", err
+	}
+
+	return userInfo.Name + "/" + clusterNick, nil
+}
+
+func getContextNickname(namespace string, clientCfg *client.Config) (string, error) {
+	client, err := osclient.New(clientCfg)
+	if err != nil {
+		return "", err
+	}
+	userInfo, err := client.Users().Get("~")
+	if err != nil {
+		return "", err
+	}
+
+	clusterNick, err := getClusterNickname(clientCfg)
+	if err != nil {
+		return "", err
+	}
+
+	return namespace + "/" + clusterNick + "/" + userInfo.Name, nil
+}
+
 // CreatePartialConfig takes a clientCfg and builds a config (kubeconfig style) from it.
-func CreateConfig(credentialsName string, namespace string, clientCfg *client.Config) clientcmdapi.Config {
+func CreateConfig(namespace string, clientCfg *client.Config) (*clientcmdapi.Config, error) {
+	clusterNick, err := getClusterNickname(clientCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	userNick, err := getUserNickname(clientCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	contextNick, err := getContextNickname(namespace, clientCfg)
```

Comment:
> should we avoid fetching the user twice from the server?

I decided to keep different methods completely distinct, figuring that the cost was minimal and not worth messing with.  Would you prefer a single method with three returns?